### PR TITLE
Add new handwired keyboard (Ergocheap)

### PR DESCRIPTION
**Pull request by xSteins**
## Description

Finalized keyboard firmware

## Types of Changes

- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/handwired/ergocheap/config.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+#pragma once
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xFEED
+#define PRODUCT_ID      0x6465
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    xSteins
+#define PRODUCT         ERGOCHEAP
+#define DESCRIPTION     ergo with bluepill
```
Comment (drashna) [inline keyboards/handwired/ergocheap/config.h:26]:
```suggestion
```

File: keyboards/handwired/ergocheap/ergocheap.c [hunk 2]
```diff
@@ -0,0 +1 @@
+#include "ergocheap.h"
```
Comment (drashna) [inline keyboards/handwired/ergocheap/ergocheap.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/handwired/ergocheap/rules.mk [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+EXTRAKEY_ENABLE = yes       # Audio control and System control(+450)
+CONSOLE_ENABLE = yes        # Console for debug(+400)
+COMMAND_ENABLE = yes         # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = yes       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality on B7 by default
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+DEFAULT_FOLDER = ergocheap
```
Comment (drashna) [inline keyboards/handwired/ergocheap/rules.mk:24]:
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/handwired/ergocheap/ergocheap.c (hunk 2 above)
Comment (xSteins) [reply keyboards/handwired/ergocheap/ergocheap.c:16]: i have updated the code

File: keyboards/handwired/ergocheap/ergocheap.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+/* 
+ * The first section contains all of the arguments representing the physical
+ * layout of the board and position of the Keys.
+ *
+ * The second converts the arguments into a two-dimensional array which
+ * represents the switch matrix.
+*/
+
+
+#define LAYOUT_65_ANSI( \
```
Comment (drashna) [inline keyboards/handwired/ergocheap/ergocheap.h:29]:
```suggestion
#define LAYOUT_65_ansi( \
```

File: keyboards/handwired/ergocheap/keymaps/default/keymap.c [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+#include QMK_KEYBOARD_H
+
+// Defines names for use in layer keycodes and the keymap
+enum layer_names {
+	_BASE,
+	_FN1,
+};
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [_BASE] = LAYOUT(
```
Comment (drashna) [inline keyboards/handwired/ergocheap/keymaps/default/keymap.c:26]: This needs to match what is in the keyboard's header file. 
```suggestion
    [_BASE] = LAYOUT_65_ansi(
```

File: keyboards/handwired/ergocheap/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+              KC_ESC,  KC_1,    KC_2,                                                                                   KC_MINS, KC_EQL,  KC_BSPC,          KC_GRV,
+            KC_TAB,       KC_Q,       KC_3,    KC_4,    KC_5,    KC_6,                 KC_7,    KC_8,    KC_9,    KC_0,          KC_LBRC, KC_RBRC, KC_BSLS,     KC_DEL,
+          KC_CAPS,        KC_A,  KC_W,    KC_E,    KC_R,    KC_T,                 KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_SCLN, KC_QUOT, KC_ENT,             KC_HOME,
+        KC_LSFT,             KC_Z, KC_S,    KC_D,    KC_F,    KC_G,                 KC_H,    KC_J,    KC_K,    KC_L,           KC_SLSH,  KC_LALT,    KC_UP,     KC_END,
+        KC_LCTL,                        KC_X,    KC_C,    KC_V,    KC_B,               KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_LGUI,         KC_LEFT, KC_DOWN, KC_RGHT,
+                                        KC_LALT,   KC_SPC,           MO(_FN1),        KC_SPC,                  MO(_FN1)
+    ),
+    [_FN1] = LAYOUT(
```
Comment (drashna) [inline keyboards/handwired/ergocheap/keymaps/default/keymap.c:34]:
```suggestion
    [_FN1] = LAYOUT_65_ansi(
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by xSteins** [comment]
Can i get this keyboard merged? i have fixed those bugs :) @drashna

**Review by drashna — APPROVED** [review]

File: keyboards/handwired/ergocheap/readme.md [hunk 7]
```diff
@@ -0,0 +1,13 @@
+# Ergocheap handwired keyboard
+
+Ergocheap is an ergonomic keyboard with layout forked from adelheid with some minor adjustment in 13-15th row and standard keycaps size
+![Ergocheap](https://i.imgur.com/II3aBYg.jpg)
+* Keyboard Maintainer: [xSteins](https://github.com/xSteins)
```
Comment (fauxpark) [inline keyboards/handwired/ergocheap/readme.md:5]:
```suggestion

![Ergocheap](https://i.imgur.com/II3aBYgl.jpg)

* Keyboard Maintainer: [xSteins](https://github.com/xSteins)
```

File: keyboards/handwired/ergocheap/rules.mk [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+MOUSEKEY_ENABLE = yes       # Mouse keys(+4700)
+EXTRAKEY_ENABLE = yes       # Audio control and System control(+450)
+CONSOLE_ENABLE = yes        # Console for debug(+400)
+COMMAND_ENABLE = yes         # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = yes       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality on B7 by default
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
```
Comment (fauxpark) [inline keyboards/handwired/ergocheap/rules.mk:23]:
```suggestion
BOOTMAGIC_ENABLE = full     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = yes        # Console for debug
COMMAND_ENABLE = yes        # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = yes      # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/handwired/ergocheap/keymaps/default/keymap.c [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    [_FN1] = LAYOUT_65_ANSI(
+              KC_GRV,  KC_F1,   KC_F2,                                                                                  KC_F11,  KC_F12,  _______,          _______,
+            _______,      _______,    KC_F3,   KC_F4,   KC_F5,   KC_F6,                KC_F7,   KC_F8,   KC_F9,   KC_F10,        _______, _______, _______,     _______,
+          KC_LCTL,        KC_LEFT, KC_UP,   _______, _______, _______,              _______, KC_MPRV, KC_MPLY, KC_MNXT, _______, _______, _______, _______,            _______,
+        KC_LSFT,          _______, KC_DOWN, KC_RGHT, _______, _______,              _______, KC_VOLD, KC_MUTE, KC_VOLU,        _______,  _______,    _______,   _______,
+        _______,                        _______, _______, _______, _______,            _______, _______, _______, _______, _______,           _______, _______, _______,
+                                        KC_LGUI,   _______,          _______,     _______,                 _______
+    )
+};
+*/
```
Comment (fauxpark) [inline keyboards/handwired/ergocheap/keymaps/default/keymap.c:61]:
```suggestion
```

File: keyboards/handwired/ergocheap/keymaps/via/keymap.c [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    [_FN1] = LAYOUT_65_ANSI(
+              KC_GRV,  KC_F1,   KC_F2,                                                                                  KC_F11,  KC_F12,  _______,          _______,
+            _______,      _______,    KC_F3,   KC_F4,   KC_F5,   KC_F6,                KC_F7,   KC_F8,   KC_F9,   KC_F10,        _______, _______, _______,     _______,
+          KC_LCTL,        KC_LEFT, KC_UP,   _______, _______, _______,              _______, KC_MPRV, KC_MPLY, KC_MNXT, _______, _______, _______, _______,            _______,
+        KC_LSFT,          _______, KC_DOWN, KC_RGHT, _______, _______,              _______, KC_VOLD, KC_MUTE, KC_VOLU,        _______,  _______,    _______,   _______,
+        _______,                        _______, _______, _______, _______,            _______, _______, _______, _______, _______,           _______, _______, _______,
+                                        KC_LGUI,   _______,          _______,     _______,                 _______
+    )
+};
+*/
```
Comment (fauxpark) [inline keyboards/handwired/ergocheap/keymaps/via/keymap.c:61]:
```suggestion
```

File: keyboards/handwired/ergocheap/keymaps/via/rules.mk [hunk 11]
```diff
@@ -0,0 +1,2 @@
+VIA_ENABLE = yes
+LTO_ENABLE = no
```
Comment (fauxpark) [inline keyboards/handwired/ergocheap/keymaps/via/rules.mk:2]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/handwired/ergocheap/keymaps/via/keymap.c [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+                 KC_LALT, KC_SPC,           MO(_FN1),           KC_SPC,           MO(_FN1)
+    ),
+    [_FN1] = LAYOUT(
+        KC_GRV,  KC_F1,   KC_F2,                                                                    KC_F11,  KC_F12,  _______, _______,
+        _______, _______, KC_F3,   KC_F4,   KC_F5,   KC_F6,     KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_UP, _______, _______, _______,
+        KC_LCTL, _______, _______, _______, _______, _______,   _______, KC_MPRV, KC_MPLY, KC_MNXT, _______, KC_LEFT, KC_RIGHT, _______, _______,
+        KC_LSFT, KC_LEFT, KC_S,    _______, _______, _______,   _______, KC_VOLD, KC_MUTE, KC_VOLU, KC_DOWN, _______, KC_MS_WH_UP,   _______,
+        _______, KC_UP,   KC_DOWN, KC_RGHT, _______,            _______, _______, _______, _______, _______, KC_MS_WH_LEFT, KC_MS_WH_DOWN, KC_MS_WH_RIGHT,
+                 KC_LGUI, _______,          _______,            _______,          _______
+    )
```
Comment (fauxpark) [inline keyboards/handwired/ergocheap/keymaps/via/keymap.c:41]: Need two more empty layers here.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/handwired/ergocheap/keymaps/via/keymap.c (hunk 12 above)
Comment (xSteins) [reply keyboards/handwired/ergocheap/keymaps/via/keymap.c:41]: i need to have 4 layers?

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by EmmStewie2709 — APPROVED** [review]

**Review by EmmStewie2709 — APPROVED** [review]

**Review by EmmStewie2709 — APPROVED** [review]

**Review by EmmStewie2709 — APPROVED** [review]

**Review by EmmStewie2709 — APPROVED** [review]

**Review by EmmStewie2709 — APPROVED** [review]

**Comment by Erovia** [comment]
Hey @xSteins, can you look into updating the ChibiOS-related config files as fauxpark mentioned?

**Comment by xSteins** [comment]
> Hey @xSteins, can you look into updating the ChibiOS-related config files as fauxpark mentioned?

Sorry for late reply, i have updated the code. @Erovia

**Review by Erovia — APPROVED** [review]

**Review by fauxpark — CHANGES_REQUESTED** [review]
chconf and halconf still need migrating.

**Comment by xSteins** [comment]
> chconf and halconf still need migrating.

done

**Review by fauxpark — APPROVED** [review]

**Pull request merged**